src/class-wc-gateway-btcpay.php

```diff
@@ -969,21 +970,28 @@ public function ipn_callback()
                     wp_die('Invalid IPN');
                 }
             } catch (\Exception $e) {
-                $error_string = 'IPN Check: Can\'t find invoice ' . $json['id'] . ' (Order ID: ' . $json['orderId'] . ')';
+                $error_string = 'IPN Check: Can\'t find invoice ' . $json['id'] . ' (Order Number: ' . $json['orderId'] . ')';
                 $this->log("    [Error] $error_string");
                 $this->log("    [Error] " . $e->getMessage());
 
                 wp_die($e->getMessage());
             }
 
-            $order_id = $invoice->getOrderId();
+            // The BTCPay order id is the WooCommerce order number, see $invoice->setOrderId in process_payment()
+            $order_number = $invoice->getOrderId();
+
+            // We get the actual WooCommerce ID from the pos data, see $invoice->setPosData in process_payment()
+            // The posData is a string and needs to be JSON decoded separately.
+            $posData = json_decode($json['posData'], true);
+            $order_id = $posData['WooCommerce']['Order ID'];
+
             $responseData = json_decode($client->getResponse()->getBody());
 
             if (false === isset($order_id) && true === empty($order_id)) {
                 $this->log('    [Error] The BTCPay payment plugin was called to process an IPN message but could not obtain the order ID from the invoice.');
                 throw new \Exception('The BTCPay payment plugin was called to process an IPN message but could not obtain the order ID from the invoice. Cannot continue!');
             } else {
-                $this->log('    [Info] Order ID is: ' . $order_id);
+                $this->log('    [Info] Order ID is: ' . $order_id. ' (Order Number: ' . $order_number . ')');
             }
 
             $order = wc_get_order($order_id);
```

Comment:
Discussion point (no action needed): Not sure how data is stored in WP but in Drupal I store the BTCPay invoice ID also as metadata alongside the order. To avoid using posData we could load the order by the BTCPay invoice ID here with a helper function and avoid all of the above (e.g. `wc_get_btcpay_order($json['id'])`) But maybe this is something for the major rewrite/new development for the greenfield API.

Reply:
Yes, we actually store some [BTCPay meta data](https://github.com/btcpayserver/woocommerce-plugin/blob/master/src/class-wc-gateway-btcpay.php#L827) with the invoice. This meta data is stores with the WooCommeerce order id as a key. In the `ipn_callback()` method we need to first get this order ID – hence it neeeds to be part of the data coming in via the POST. Otherwise we don't know the key for the lookup.

Reply:
@dennisreimann ok so the BTCPay invoice ID gets stored alongside the order where you mentioned it like this:
>  update_post_meta($order_id, 'BTCPay_id', $invoice->getId());

So we could use meta query to load the order by the BTCPay invoice id and avoid handling order id vs order number alltogether: https://rudrastyh.com/wordpress/meta_query.html or get_posts() https://developer.wordpress.org/reference/functions/get_posts/#comment-1196

e.g.: (did not test, just example);
```
$args = array(
    'post_type'  => 'order',
    'meta_query' => array(
        array(
            'key'   => 'BTCPay_id',
            'value' => $json['id'],
        )
    )
);
$order = get_posts( $args );
```

Reply:
Added in 27a383b3643fdc58066e659d0b212af24ab8ad61. Tested with sample data, but I'm waiting for an actual order to confirm it works.

Reply:
@dennisreimann This opens us to a number of edge-cases  now. It would have been better to discuss this radical shift of matching in a separate issue/PR.  
Some issues I see:
* User clicks on Pay order >1 times and generates multiple invoices. Do they both get stored in woocommerce in the same order?
* User does not pay invoice on time, will the user be able to generate a new invoice to pay?
* User generates a new invoice after old one expires but pays old invoice. Can it still be matched in woocommerce?

Reply:
Ah nvm, you seem to have a fallback method. Nevertheless, have you considered my above cases?

Reply:
@kukks I'm not sure about your concerns, because in general nothing has changed – the recent addition and the fallback are  related to how the order is matched/looked up in the `ipn_callback` method.

Besides switching back to the order number as identifier in BTCPay, the `process_payment` method is mostly the same as before.